File: backend/app/routers/preferences.py
"""
偏好引擎 API 路由
"""
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.user import User
from app.schemas.preference import (
    CustomTagCreateRequest,
    CustomTagResponse,
    CustomTagUpdateRequest,
    PreferenceDetailListResponse,
    PreferenceListResponse,
    TagChoiceListResponse,
)
from app.services.auth import get_current_user
from app.services.ledger import get_ledger_or_404, require_read_ledger, require_write_ledger
from app.services.preference import (
    create_or_restore_custom_tag,
    get_custom_tag_or_404,
    get_item_choices,
    get_location_choices,
    get_sorted_categories,
    get_sorted_items,
    get_sorted_locations,
    get_sorted_subjects,
    get_subject_preference_details,
    rename_custom_tag,
)

router = APIRouter(prefix="/ledgers/{ledger_id}/preferences", tags=["preferences"])


@router.get("/subjects", response_model=PreferenceListResponse)
async def subjects(
    ledger_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PreferenceListResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_read_ledger(db, ledger, current_user)
    return PreferenceListResponse(
        items=await get_sorted_subjects(db, ledger_id, current_user.id)
    )


@router.get("/subjects/details", response_model=PreferenceDetailListResponse)
async def subject_details(
    ledger_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PreferenceDetailListResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_read_ledger(db, ledger, current_user)
    return PreferenceDetailListResponse(
        items=await get_subject_preference_details(db, ledger_id, current_user.id)
    )


@router.get("/categories", response_model=PreferenceListResponse)
async def categories(
    ledger_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PreferenceListResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_read_ledger(db, ledger, current_user)
    return PreferenceListResponse(
        items=await get_sorted_categories(db, ledger_id, current_user.id)
    )


@router.get("/items", response_model=PreferenceListResponse)
async def items(
    ledger_id: uuid.UUID,
    category: str = Query(..., min_length=1, max_length=50),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PreferenceListResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_read_ledger(db, ledger, current_user)
    return PreferenceListResponse(
        items=await get_sorted_items(db, ledger_id, current_user.id, category)
    )


@router.get("/locations", response_model=PreferenceListResponse)
async def locations(
    ledger_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PreferenceListResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_read_ledger(db, ledger, current_user)
    return PreferenceListResponse(
        items=await get_sorted_locations(db, ledger_id, current_user.id)
    )


@router.get("/items/details", response_model=TagChoiceListResponse)
async def item_details(
    ledger_id: uuid.UUID,
    category: str = Query(..., min_length=1, max_length=50),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TagChoiceListResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_read_ledger(db, ledger, current_user)
    return TagChoiceListResponse(items=await get_item_choices(db, ledger_id, current_user.id, category))


@router.get("/locations/details", response_model=TagChoiceListResponse)
async def location_details(
    ledger_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TagChoiceListResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_read_ledger(db, ledger, current_user)
    return TagChoiceListResponse(items=await get_location_choices(db, ledger_id, current_user.id))


def _custom_tag_response(tag) -> CustomTagResponse:
    return CustomTagResponse(
        id=tag.id,
        tag_type=tag.tag_type,
        name=tag.name,
        category=tag.scope or None,
        is_hidden=tag.is_hidden,
    )


@router.post("/tags", response_model=CustomTagResponse)
async def create_custom_tag(
    ledger_id: uuid.UUID,
    payload: CustomTagCreateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CustomTagResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_write_ledger(db, ledger, current_user)
    tag = await create_or_restore_custom_tag(
        db,
        ledger_id,
        payload.tag_type,
        payload.name.strip(),
        payload.category,
    )
    return _custom_tag_response(tag)


@router.patch("/tags/{tag_id}", response_model=CustomTagResponse)
async def update_custom_tag(
    ledger_id: uuid.UUID,
    tag_id: uuid.UUID,
    payload: CustomTagUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CustomTagResponse:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_write_ledger(db, ledger, current_user)
    tag = await get_custom_tag_or_404(db, ledger_id, tag_id)
    return _custom_tag_response(await rename_custom_tag(db, tag, payload.name.strip()))


@router.delete("/tags/{tag_id}")
async def hide_custom_tag(
    ledger_id: uuid.UUID,
    tag_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, str]:
    ledger = await get_ledger_or_404(db, ledger_id)
    await require_write_ledger(db, ledger, current_user)
    tag = await get_custom_tag_or_404(db, ledger_id, tag_id)
    tag.is_hidden = True
    await db.flush()
    return {"detail": "Custom tag hidden successfully."}

File: frontend/src/api/preferences.ts
import { apiClient } from '@/api'

export interface PreferenceListResponse {
  items: string[]
}

export interface PreferenceItem {
  value: string
  selection_count: number
  last_selected_at: string | null
}

export interface PreferenceDetailListResponse {
  items: PreferenceItem[]
}

export interface Subject {
  id: string
  ledger_id?: string
  name: string
  is_preset?: boolean
  is_hidden?: boolean
  display_order?: number
  created_at?: string
}

export interface TagChoice {
  id: string | null
  value: string
  is_system: boolean
  selection_count: number
  last_selected_at: string | null
}

export interface CustomTag {
  id: string
  tag_type: 'item' | 'location'
  name: string
  category: string | null
  is_hidden: boolean
}

export async function getPreferredCategories(ledgerId: string): Promise<string[]> {
  const response = await apiClient.get<PreferenceListResponse>(`/ledgers/${ledgerId}/preferences/categories`)
  return response.data.items
}

export async function getPreferredItems(ledgerId: string, category: string): Promise<string[]> {
  const response = await apiClient.get<PreferenceListResponse>(`/ledgers/${ledgerId}/preferences/items`, {
    params: { category },
  })
  return response.data.items
}

export async function getItemChoices(ledgerId: string, category: string): Promise<TagChoice[]> {
  const response = await apiClient.get<{ items: TagChoice[] }>(`/ledgers/${ledgerId}/preferences/items/details`, {
    params: { category },
  })
  return response.data.items
}

export async function getPreferredLocations(ledgerId: string): Promise<string[]> {
  const response = await apiClient.get<PreferenceListResponse>(`/ledgers/${ledgerId}/preferences/locations`)
  return response.data.items
}

export async function getLocationChoices(ledgerId: string): Promise<TagChoice[]> {
  const response = await apiClient.get<{ items: TagChoice[] }>(`/ledgers/${ledgerId}/preferences/locations/details`)
  return response.data.items
}

export async function getPreferredSubjects(ledgerId: string): Promise<string[]> {
  const response = await apiClient.get<PreferenceListResponse>(`/ledgers/${ledgerId}/preferences/subjects`)
  return response.data.items
}

export async function getSubjectPreferenceDetails(ledgerId: string): Promise<PreferenceItem[]> {
  const response = await apiClient.get<PreferenceDetailListResponse>(`/ledgers/${ledgerId}/preferences/subjects/details`)
  return response.data.items
}

export async function listSubjects(ledgerId: string): Promise<Subject[]> {
  const response = await apiClient.get<Subject[]>(`/ledgers/${ledgerId}/subjects`)
  return response.data
}

export async function createSubject(ledgerId: string, name: string): Promise<Subject> {
  const response = await apiClient.post<Subject>(`/ledgers/${ledgerId}/subjects`, { name })
  return response.data
}

export async function deleteSubject(ledgerId: string, subjectId: string): Promise<void> {
  await apiClient.delete(`/ledgers/${ledgerId}/subjects/${subjectId}`)
}

export async function updateSubject(ledgerId: string, subjectId: string, name: string): Promise<Subject> {
  const response = await apiClient.patch<Subject>(`/ledgers/${ledgerId}/subjects/${subjectId}`, { name })
  return response.data
}

export async function createCustomTag(
  ledgerId: string,
  payload: { tag_type: 'item' | 'location'; name: string; category?: string },
): Promise<CustomTag> {
  const response = await apiClient.post<CustomTag>(`/ledgers/${ledgerId}/preferences/tags`, payload)
  return response.data
}

export async function updateCustomTag(ledgerId: string, tagId: string, name: string): Promise<CustomTag> {
  const response = await apiClient.patch<CustomTag>(`/ledgers/${ledgerId}/preferences/tags/${tagId}`, { name })
  return response.data
}

export async function deleteCustomTag(ledgerId: string, tagId: string): Promise<void> {
  await apiClient.delete(`/ledgers/${ledgerId}/preferences/tags/${tagId}`)
}
